File: string/1371_find_the_longest_substring_containing_vowels_in_even_counts/1371_find_the_longest_substring_containing_vowels_in_even_counts.go
package main

import (
	"fmt"
	"math"
)

var target = map[uint8]int{
	'a': 1,
	'e': 2,
	'i': 4,
	'o': 8,
	'u': 16,
}

func findTheLongestSubstring(s string) int {
	var (
		i, curr, rst, st int
		ok               bool
		length           = len(s)
		status           = make([]int, 32)
	)
	for i = 0; i < 32; i++ {
		status[i] = math.MaxInt32
	}
	status[0] = -1
	for i = 0; i < length; i++ {
		if st, ok = target[s[i]]; ok {
			curr ^= st
		}
		if status[curr] == math.MaxInt32 {
			status[curr] = i
		} else {
			rst = max(rst, i-status[curr])
		}
	}
	return rst
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
func main() {
	s := "eleetminicoworoep"
	res := findTheLongestSubstring(s)
	fmt.Println(res)
}

File: string/1371_find_the_longest_substring_containing_vowels_in_even_counts/1371_find_the_longest_substring_containing_vowels_in_even_counts.py

class Solution(object):
    def findTheLongestSubstring(self, s):
        vowels = {'a':1, 'e':2, 'i':4, 'o':8, 'u':16}
        d, n , r = {0:-1}, 0, 0
        for i, c in enumerate(s):
            if c in vowels:
                n ^= vowels[c]
            if n not in d:
                d[n] = i
            else:
                r = max(r, i - d[n])
            # print c, bin(n), d, r
        return r


s = "eleetminicoworoep"
res = Solution().findTheLongestSubstring(s)
print(res)